src/emulator/renderer/src/vulkan/renderer.cpp

```diff
@@ -0,0 +1,574 @@
+// Vita3K emulator project
+// Copyright (C) 2018 Vita3K team
+//
+// This program is free software; you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation; either version 2 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License along
+// with this program; if not, write to the Free Software Foundation, Inc.,
+// 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+
+#include <renderer/functions.h>
+#include <renderer/types.h>
+
+#include <config/version.h>
+#include <features/state.h>
+
+#include <renderer/vulkan/functions.h>
+#include <renderer/vulkan/state.h>
+
+#include <SDL.h>
+#include <SDL_vulkan.h>
+
+#include <util/log.h>
+
+// Setting a default value for now.
+// In the future, it might be a good idea to take the host's device memory into account.
+constexpr static size_t private_allocation_size = MB(1);
+
+// Some random number as value. It will likely be very different. There are probably SCE feilds for this, I will look later.
+constexpr static uint32_t max_sets = 192;
+constexpr static uint32_t max_buffers = 64;
+constexpr static uint32_t max_images = 64;
+constexpr static uint32_t max_samplers = 64;
+
+constexpr static vk::Format screen_format = vk::Format::eB8G8R8A8Unorm;
+
+const static std::vector<const char *> instance_layers = {
+#ifndef NDEBUG
+    "VK_LAYER_LUNARG_standard_validation",
+#endif
+};
+
+const static std::vector<const char *> instance_extensions = {
+    "VK_KHR_surface",
+#ifdef __APPLE__
+    "VK_MVK_macos_surface",
```

Comment:
Just a small recommendation but `VK_MVK_macos_surface` has been depreciated in favour of `VK_EXT_metal_surface` as stated in the MoltenVK runtime [guide](https://github.com/google/angle/blob/master/src/libANGLE/renderer/d3d/d3d11/Renderer11.cpp) so prefer to use that extension instead.

Reply:
Ah I had no idea! Thanks a lot. I think SDL threw an error message earlier asking me to activate this one. It also doesn't appear in my `vulkaninfo` listing... Let me try.

Reply:
You are [absolutely right](https://github.com/KhronosGroup/MoltenVK/blob/master/Docs/MoltenVK_Runtime_UserGuide.md)! But I don't seem to have `VK_EXT_metal_surface`... Do I have to enable it or do something special? My Vulkan SDK version happens to be 1.1.108 and I snatched it from LunarG distribution site.

Reply:
Try grabbing the latest version of MoltenVK from the repository and make sure your macOS device is up to date as well ?

Reply:
Yeap, it looks like they only added just [last](https://github.com/KhronosGroup/MoltenVK/commit/aa6a10570a4ce0077120d49dd24c70fd5f5d8481#diff-82d68bf9683270d8d20a27e6da6ad9a3) month in MoltenVK.

Reply:
EH! I'm actually 3 updates behind at macOS 10.14.3- that's probably why. I've been holding off updating because 10.14.4 SDK broke OpenGL support and it would stop me from working on this project. I have no idea if it was fixed in .5 or .6.

Just wondering, if I do update and I only use VK_EXT_metal_surface, does that mean Vita3K Vulkan will drop support for macOS versions earlier than Mojave?

Reply:
Oh- or not. Maybe its just an SDK thing. Phew.

Reply:
Generally, it depends on what extensions you're going to need that'll bump up the macOS version requirements. For an example, if you're going to require fragment shader interlock which is only available in Metal 2 then your app will need macOS Mojave at minimum with devices that also support that extension as well. 

Apple has a good track record of supporting older devices on Metal anyways so any devices that worked with the initial release of Metal will also work in the upcoming latest Metal 3.0 which will be available in macOS Catalina. If I were a MoltenVK user then I would just upgrade without worrying too much for now since even Metal 3.0 will give you just about everything you would want. 

Hypothetically, as a macOS developer I'd be more concerned if my device supported [raster order groups](https://developer.apple.com/documentation/metal/gpu_features/understanding_gpu_family_4/about_raster_order_groups) or not since it's only available on AMD Vega/Navi GPUs and it'll have a profound impact on emulation as well.

Reply:
Sorry to bother again but here are some notes below on how to enable `VK_EXT_metal_surface`.

> You can enable the `VK_EXT_metal_surface` extension by defining the `VK_USE_PLATFORM_METAL_EXT` guard macro in your compiler build settings. See the description of the `mvk_vulkan.h` file below for a convenient way to enable this extension automatically.

> The `mvk_vulkan.h` header file automatically enables the `VK_USE_PLATFORM_METAL_EXT` 
   build setting and `VK_EXT_metal_surface` *Vulkan* extension

.Did you include the MoltenVK header file or use the guard macro in your project yet because I don't see them in this PR ?

Reply:
I can guarantee you that I absolutely did not do that. I'll do that in a bit. I made an a special function that checks if you have it in your extension list though :/

I'll include it. Thanks for the heads up.